### PR TITLE
New simple package to provide exponential backoff

### PR DESCRIPTION
**Pull request by tgraf**
Dependency for #2613

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: pkg/backoff/backoff.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/cilium/cilium/pkg/uuid"
+
+	"github.com/sirupsen/logrus"
+)
+
+var log = logging.DefaultLogger
+
+// Exponential implements an exponential backoff
+type Exponential struct {
+	Min    time.Duration
```
Comment (aanm) [inline pkg/backoff/backoff.go]: Can you add comments describing what is the purpose of each field?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/backoff/backoff.go (hunk 1 above)
Comment (tgraf) [reply pkg/backoff/backoff.go]: fixed

**Comment by raybejjani** [comment]
Note: This is a drive-by comment, you can ignore it and merge.
I'm not sure how many use-cases you envision. Could you also have a function that returns a timer channel? This would be useful in goroutines that might be waiting on multiple things (like a retry backoff and a new-event trigger). In a similar vein, a function that returned the the deadline time to be used with SetDeadline might be useful, but I can't actually think of a case where I would use it.

**Comment by nebril** [comment]
Code LGTM, although @raybejjani's comment makes sense for me.

I am not fond of adding this without any usage, at this point this is just dead code added to repo.

**Comment by tgraf** [comment]
> I'm not sure how many use-cases you envision. Could you also have a function that returns a timer channel? This would be useful in goroutines that might be waiting on multiple things (like a retry backoff and a new-event trigger). In a similar vein, a function that returned the the deadline time to be used with SetDeadline might be useful, but I can't actually think of a case where I would use it.

Agreed. I kept this minimal as per dependency for #2613. We can easily extend this to also include a channel later on.

**Comment by tgraf** [comment]
> I am not fond of adding this without any usage, at this point this is just dead code added to repo.

#2613 contains the code which uses this. I wanted to factor it out to simplify review.

Usage it trivial:
``` go
                backoffTemplate: backoff.Exponential{
                        Min:    time.Duration(20) * time.Millisecond,
                        Factor: 2.0,
                },


[...]
        // make a copy of the template and customize it
        boff := a.backoffTemplate
        boff.Name = key.String()

         for {
                boff.Wait()
         }
```

**Pull request merged**